src/cockpit/protocol.py

```diff
@@ -131,15 +131,19 @@ def connection_lost(self, exc):
     def write_frame(self, frame):
         frame_length = len(frame)
         header = f'{frame_length}\n'.encode('ascii')
-        self.transport.write(header + frame)
+        if self.transport is not None:
```

Comment:
Should this raise a warning or even assertion? This doesn't just look like an "remote peer unexpectedly closed its end" situation, as `self.transport` is local state -- i.e. this smells like a logic error in our code? I.e. why do we still call write_frame() after getting connection_lost()?

Reply:
This isn't about the peer but rather about the main stdio transport.  If the connection closes suddenly due to a protocol error or something and we're in the middle of writing a bunch of stuff from some channel, this is the case where we try to write these frames.

It's annoying too, since the original error message gets lost in s bunch of unrelated tracebacks about "NoneType has no write method".

I think that asyncio.run() is partially responsible here too.  We return from the main() function in the bridge as soon as we see that the connection dropped, but it seems to want to run outstanding tasks to completion.   Or at least there are probably cases where some tasks have IO scheduled before our communicate() future callback gets run.